tests/ostree-spawn.c

```diff
@@ -0,0 +1,341 @@
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "misc-utils.h"
+#include "ostree-spawn.h"
+
+#include <errno.h>
+
+#ifndef OSTREE_BINARY
+#error OSTREE_BINARY is not defined
+#endif
+
+static void
+copy_strv_to_ptr_array (gchar **strv,
+			GPtrArray *array)
+{
+  gchar **iter;
+
+  if (strv != NULL)
+    for (iter = strv; *iter != NULL; ++iter)
+      g_ptr_array_add (array, g_strdup (*iter));
+  g_ptr_array_add (array, NULL);
+}
+
+static gboolean
+spawn_ostree_in_repo (GFile *repo,
+		      gchar **args,
+		      CmdStuff *cmd,
+		      GError **error)
+{
+  SSDEF;
+  g_autoptr(GPtrArray) argv = string_array_new ();
+  g_auto(GStrv) raw_argv = NULL;
+
+  g_ptr_array_add (argv, g_strdup (OSTREE_BINARY));
+  g_ptr_array_add (argv, flag ("repo", SS (g_file_get_path (repo))));
+  copy_strv_to_ptr_array (args, argv);
+  raw_argv = (gchar**)g_ptr_array_free (g_steal_pointer (&argv), FALSE);
+
+  return test_spawn (raw_argv, NULL, cmd, error);
+}
+
+static const gchar *
+repo_mode_to_string (RepoMode mode)
+{
+  switch (mode)
+    {
+    case REPO_ARCHIVE_Z2:
+      return "archive-z2";
+
+    case REPO_BARE:
+      return "bare";
+    }
+
+  g_assert_not_reached ();
+}
+
+gboolean
+ostree_init (GFile *repo,
+             RepoMode mode,
+             CmdStuff *cmd,
+             GError **error)
+{
+  SSDEF;
+  g_auto(GStrv) args = NULL;
+
+  args = generate_strv ("init",
+                        SS (flag ("mode", repo_mode_to_string (mode))),
+                        NULL);
+
+  return spawn_ostree_in_repo (repo,
+			       args,
+			       cmd,
+			       error);
+}
+
+gboolean
+ostree_commit (GFile *repo,
+               GFile *tree_root,
+               const gchar *subject,
+               const gchar *ref,
+               const gchar *keyid,
+               GDateTime *timestamp,
+               CmdStuff *cmd,
+               GError **error)
+{
+  SSDEF;
+  g_auto(GStrv) args = NULL;
+
+  args = generate_strv ("commit",
+                        SS (flag ("subject", subject)),
+                        SS (flag ("branch", ref)),
+                        SS (flag ("gpg-sign", keyid)),
+                        "--gpg-homedir=" GPG_HOME_DIRECTORY,
+                        SS (flag ("timestamp", SS (g_date_time_format (timestamp, "%F")))),
+                        SS (g_file_get_path (tree_root)),
+                        NULL);
```

Comment:
This is clever and terse, however ....

I wonder if it would be better to have a helper function here like `build_cmdline_args` which either just takes an array of something like `struct { const char *name, const char *value }` (or varargs) and creates the command line arguments from there. That way we don't need to muck about with this `SSDEF` stuff either.